ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,507 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+            AmountToReturn = Amount;
+        }
+
+        /* Return the amounting we have back to Memory Manager */
+        InterlockedExchangeAddSizeT(QuotaBlock->QuotaEntry[QuotaType].Return, AmountToReturn);
+        MmReturnPoolQuota(QuotaType, QuotaBlock->QuotaEntry[QuotaType].Return);
+
+        /* And try to trim the limit */
+        InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Limit,
+                                        QuotaBlock->QuotaEntry[QuotaType].Limit - AmountToReturn,
+                                        QuotaBlock->QuotaEntry[QuotaType].Limit);
     }
 
+    /* Update the usage member of the block */
+    UpdatedQuotaUsage = QuotaBlock->QuotaEntry[QuotaType].Usage - Amount;
+    InterlockedCompareExchangeSizeT(&QuotaBlock->QuotaEntry[QuotaType].Usage,
+                                    UpdatedQuotaUsage,
+                                    QuotaBlock->QuotaEntry[QuotaType].Usage);
+
+    /* Are we being given a process? */
+    if (Process)
+    {
+        /* We're being given, check that's not a system one  */
+        ASSERT(Process != PsInitialSystemProcess);
+
+        /* Decrease the process' quota usage */
+        InterlockedExchangeAddSizeT(&Process->QuotaUsage[QuotaType], -(LONG)Amount);
+    }
+
+    /* We're done, release the lock */
     KeReleaseSpinLock(&PspQuotaLock, OldIrql);
 }
 
 /* FUNCTIONS ***************************************************************/
 
+/**
+ * @brief
+ * Initializes the quota system during boot
+ * phase of the system, which sets up the
+ * default quota block that is used across
+ * several processes.
+ *
+ * @return
+ * Nothing.
+ */
 CODE_SEG("INIT")
 VOID
 NTAPI
 PsInitializeQuotaSystem(VOID)
 {
+    /* Zero out the block */
     RtlZeroMemory(&PspDefaultQuotaBlock, sizeof(PspDefaultQuotaBlock));
+
+    /* Assign the default quota limits */
     PspDefaultQuotaBlock.QuotaEntry[PsNonPagedPool].Limit = (SIZE_T)-1;
     PspDefaultQuotaBlock.QuotaEntry[PsPagedPool].Limit = (SIZE_T)-1;
     PspDefaultQuotaBlock.QuotaEntry[PsPageFile].Limit = (SIZE_T)-1;
+
+    /*
+     * Set up the count references as the
+     * default block will gonna be used.
+     */
+    PspDefaultQuotaBlock.ReferenceCount = 1;
+    PspDefaultQuotaBlock.ProcessCount = 1;
+
+    /* Assign that block to initial process */
     PsGetCurrentProcess()->QuotaBlock = &PspDefaultQuotaBlock;
 }
 
+/**
+ * @brief
+ * Inherits the quota block to another newborn
+ * (child) process. If there's no parent
+ * process, the default quota block is
+ * assigned.
+ *
+ * @param[in] Process
+ * The child process which quota block
+ * is to be given.
+ *
+ * @param[in] ParentProcess
+ * The parent process.
+ *
+ * @return
+ * Nothing.
+ */
 VOID
 NTAPI
-PspInheritQuota(PEPROCESS Process, PEPROCESS ParentProcess)
+PspInheritQuota(
+    _In_ PEPROCESS Process,
+    _In_opt_ PEPROCESS ParentProcess)
 {
+    PEPROCESS_QUOTA_BLOCK ParentQuotaBlock, DefaultQuotaBlock;
```

Comment:
It doesn't appear necessary to have two separate variables with single-usage use, so to speak.
The original `PEPROCESS_QUOTA_BLOCK QuotaBlock` was fine to be used in the two independent `if` cases.